### PR TITLE
Remove block initialisation from `caml_obj_block`

### PR DESCRIPTION
**Pull request by kayceesrk**
`caml_alloc` used to allocate the object already initialises objects with tag < `No_scan_tag`. Hence, the reinitialisation is unnecessary.

The use case I have in mind is to permit parallel initialization of an array of integers in Multicore OCaml. The idea is to allocate the array using `Obj.new_block` with `Abstract_tag` to prevent it from being traced by the GC, followed by `Obj.obj` on the result to cast it to an `int array`. This array can now be safely initialized in parallel. For an array of floats, we can already do that thanks to `Array.create_float`.

### CODE REVIEW

**Review by dra27 — APPROVED** [review]
It's been a valid optimisation for a while - it could have been applied in 1998 when `caml_alloc_small` was added (see 3be947947e3249a9b362fc790d377e43c4108a62!)

Someone else should double-check, but this looks fine to me

**Comment by dra27** [comment]
(Feel free to add a Changes entry if you want, but I don't see that it strictly needs one!)

**Comment by dra27** [comment]
Actually, on a further look, the function is basically equivalent to `return caml_alloc(Long_val(size), Long_val(tg));`

**Comment by xavierleroy** [comment]
Stop!  I agree that if tag < No_scan_tag, the initialization in `caml_obj_block` is redundant with that done by `caml_alloc` and should be removed.  However,  if tag >= No_scan_tag, the initialization is not redundant and we should at least discuss whether it's OK before approving.

**Comment by mshinwell** [comment]
I don't have a concrete example of what this might break, but I would be wary of removing the zero-initialisation, in the `>= No_scan_tag` case.  In general it seems better to return initialised memory than uninitialised memory.  I suppose a separate function could be added to `Obj` to elide the initialisation?

@kayceesrk It's unfortunate that if you do this, then you'll end up with arrays that have the wrong tag (preventing e.g.`libmonda` from being able to print such values unless it knew the type from the `.cmt` file).  Furthermore, any optimiser (e.g. Flambda 2.0) that tracks the length of arrays won't be able to know the length.  Maybe there is an easier solution?

**Comment by kayceesrk** [comment]
I agree with the concerns raised regarding returning uninitialized values. I'm all ears for an easier solution. Adding a dedicated `Int_array_tag` is ugly.

**Comment by mshinwell** [comment]
I'm not sure a dedicated tag for immediate arrays is so bad.  The issue of the GC spending a lot of time scanning such arrays unnecessarily has come up multiple times in the past.

**Comment by kayceesrk** [comment]
Focussing on this PR, in order for this optimization to be unsafe, the code allocating the new block with tag >= `No_scan_tag (251)` should rely on the fields being `Val_long(0) == 1`. The tags are:

* `Custom_tag (255)`
* `Double_array_tag (254)`
* `Double_tag (253)`
* `String_tag (252)`
* `Abstract_tag (251)`

Is it reasonable to expect the caller of `Obj.new_block` with tag >= `No_scan_tag` to rely on this fact?

**Comment by mshinwell** [comment]
It does seem like clients using `Abstract_tag` or `Custom_tag` might be the only ones.  It is also a bit peculiar since the initialiser is actually 1 and not 0.  I'm not sure the question should be whether it's reasonable, more like how much code we might break by changing this.

**Comment by kayceesrk** [comment]
I found two uses of `Obj.new_tag Obj.abstract_tag` on github. 

1. Used to create a distinguished `none` object. The contents are never examined.
https://github.com/avsm/mirage-duniverse/blob/a0c48c86332a34aadb26cdb0cf93747b34cb15ec/ocaml_modules/core_kernel/src/moption.ml#L6-L13

2. In the compiler to represent an empty raw backtrace. The size is 0 in this case, and hence, the content doesn't exist: https://github.com/ocaml/ocaml/blob/trunk/stdlib/printexc.ml#L289

I didn't find any `Obj.new_block Obj.custom_tag`.

**Comment by mshinwell** [comment]
I looked through the Jane Street tree for uses of `Obj.new_block`.  I found the following:
- One use in the `biniou` library, which contains a bug: it allocates 8 words for a value of tag `Double_tag` and only fills in the first one.
- One use in the `Option_array` library in `Base`.  This allocates a block of tag `Abstract_tag` with one field, to provide a pointer that is guaranteed to be different from all other values.  The field is left uninitialised.
- One use in the `ppx_hash` library, similarly.
- One use in the `moption` code, similarly (as @kayceesrk discovered).
- One use in `js_of_ocaml`, although it is only providing a wrapper; there is no actual useful call site.

So this change will mean that some of these values contain uninitialised data, but it would not be read in any of these cases.

**Comment by lpw25** [comment]
> It's unfortunate that if you do this, then you'll end up with arrays that have the wrong tag

Note that such arrays will not behave correctly with the non-parametric operations. So polymorphic equality, hashing and marshaling of these arrays will not work correctly.

**Comment by kayceesrk** [comment]
> Note that such arrays will not behave correctly with the non-parametric operations. So polymorphic equality, hashing and marshaling of these arrays will not work correctly.

These are very good observations. I’m less inclined to go with my proposed solution for parallel initialisation. Perhaps a distinguished tag for immediate arrays may be the best way to go forward. 

Either way, this optimisation is independently useful, if we don’t break lots of code.

**Comment by xavierleroy** [comment]
> in order for this optimization to be unsafe, the code allocating the new block with tag >= No_scan_tag (251) should rely on the fields being Val_long(0) == 1.

A slightly more general question is: what can go wrong if the contents of the block are left uninitialized?

*  Abstract_tag (251): initial contents seem irrelevant.  The uses of this tag mentioned above are somewhat superstitious: if all is needed is an block with a unique address, `Obj.repr (ref 0)` would work just as well.

* String_tag (252): on the one hand, `Bytes.create` already returns uninitialized blocks with that tag.  However, the last byte that encodes the length is properly initialized.  With the proposed change, `Obj.new_block 252 1` produces a string whose length may well be negative!

*  Double_tag (253): OK to have random garbage in the data.  (Wrong length could cause weird behaviors, but that's orthogonal to the question of initial contents.)

* Double_array_tag  (254): `Float.Array.create` already returns float arrays with garbage contents.  Should be OK.

* Custom_tag (255): this is the most problematic case, both with initialization to 1 or with no initialization.  In both cases, the first field is not a valid pointer to a `struct custom_operations`.  Hence the GC can crash trying to finalize, and ad-hoc polymorphic functions can crash trying to call custom operations.  

In summary: the current situation is already pretty bad w.r.t. well-formedness of the blocks created with `Obj.new_block`; the proposed change make it slightly worse in the case of strings, with the negative length.

**Comment by lpw25** [comment]
> This allocates a block of tag Abstract_tag with one field, to provide a pointer that is guaranteed to be different from all other values.

For the record this block has abstract tag to ensure that it fails if someone tries to marshal it, since marshalling will not work for such representations.

**Comment by kayceesrk** [comment]
> String_tag (252): on the one hand, Bytes.create already returns uninitialized blocks with that tag. However, the last byte that encodes the length is properly initialized. With the proposed change, Obj.new_block 252 1 produces a string whose length may well be negative!

We could specialize for the String_tag and set the last byte to 1 to retain the current behavior.

> Custom_tag (255): this is the most problematic case, both with initialization to 1 or with no initialization. In both cases, the first field is not a valid pointer to a struct custom_operations. Hence the GC can crash trying to finalize, and ad-hoc polymorphic functions can crash trying to call custom operations. 

We could initialize the first field to a default block of custom operations that will raise `Invalid_argument "Uninitialized custom block"`. This would prevent crashes.

**Comment by kayceesrk** [comment]
> We could initialize the first field to a default block of custom operations that will raise Invalid_argument "Uninitialized custom block". This would prevent crashes.

The manual instructs not to perform any allocation in the operations attached to custom blocks. But raising `Invalid_argument` exception will allocate. We could `caml_fatal_error` here.

**Comment by kayceesrk** [comment]
> We could initialize the first field to a default block of custom operations that will raise Invalid_argument "Uninitialized custom block". This would prevent crashes.

It appears that the runtime already handles the cases where the custom operations are NULL. We only need to initialize the first field to a block that has the default values for the operations.

**Comment by kayceesrk** [comment]
Another problem with the current implementation is that it permits `0` length string and custom blocks. 

```ocaml
$ ocaml
        OCaml version 4.09.0

# Bytes.length (Obj.obj (Obj.new_block 252 0));;
- : int = -1
# Marshal.to_bytes (Obj.obj (Obj.new_block 255 0));;
Segmentation fault (core dumped)
```

Zero length string and custom blocks should be disallowed.

**Comment by kayceesrk** [comment]
I've pushed new commits which 

* prohibits zero-length string and custom blocks
* initializes custom blocks with default operations so that operations (hash, serialize, compare, finalize) do not crash. 
* initializes strings such that the length of the string is `size of block (in bytes) - 1` (preserves the current behavior).

File: runtime/obj.c [hunk 1]
```diff
@@ -80,20 +81,39 @@ CAMLprim value caml_obj_make_forward (value blk, value fwd)
...
   tg = Long_val(tag);
-  if (sz == 0) return Atom(tg);
-  res = caml_alloc(sz, tg);
-  for (i = 0; i < sz; i++)
-    Field(res, i) = Val_long(0);
+  if (sz == 0) {
+    res = Atom(tg);
+  } else {
+    res = caml_alloc(sz, tg);
+  }
```
Comment (dra27) [inline runtime/obj.c]: Is there any value in hoisting the `sz == 0` case?
```suggestion
  caml_alloc(sz, tg)
```

File: runtime/obj.c [hunk 2]
```diff
@@ -80,20 +81,39 @@ CAMLprim value caml_obj_make_forward (value blk, value fwd)
...
```
Comment (dra27) [inline runtime/obj.c:116]: Would it be worth putting the justifications for initialisation or not for each of the tags here?

File: runtime/obj.c (hunk 1 above)
Comment (dra27) [reply runtime/obj.c]: (if not, there should really be a `CAMLassert (tg < 256)` somewhere)

Comment (kayceesrk) [reply runtime/obj.c]: Fixed. The call to `caml_alloc` correctly handles the `0` sized allocations and has the necessary assertions.

File: runtime/obj.c (hunk 2 above)
Comment (kayceesrk) [reply runtime/obj.c:116]: Added comments justifying selective initialization.

**Comment by kayceesrk** [comment]
I've added comments in the code justifying selective initialization.

File: runtime/obj.c [hunk 3]
```diff
@@ -80,19 +81,54 @@ CAMLprim value caml_obj_make_forward (value blk, value fwd)
...
+   * reason, [Custom_tag] blocks cannot be of size [0]. We initialise the first
+   * field to a block containing default custom operations.
+   */
   res = caml_alloc(sz, tg);
-  for (i = 0; i < sz; i++)
-    Field(res, i) = Val_long(0);
+
+  if (tg == String_tag) {
+    if (sz == 0) caml_invalid_argument ("Obj.new_block");
+    Field (res, sz - 1) = 1;
```
Comment (xavierleroy) [inline runtime/obj.c]: I know this is what the previous implementation of `caml_obj_block` did, and I agree that it produces a well-formed string both on little-endian and big-endian platforms.  Note however that the length of the resulting string depends on the endianness of the platform: it will be one less on a big-endian platform.  Initializing the last word with 0 instead of 1 would avoid this oddity -- and still be compatible with the current implementation on little-endian platforms, meaning almost all the platforms supported by OCaml these days.

File: runtime/obj.c [hunk 4]
```diff
@@ -80,19 +81,54 @@ CAMLprim value caml_obj_make_forward (value blk, value fwd)
...
+  } else if (tg == Custom_tag) {
+    if (sz == 0) caml_invalid_argument ("Obj.new_block");
+    Custom_ops_val (res) = &default_custom_ops;
```
Comment (xavierleroy) [inline runtime/obj.c]: Thanks for the proper initializations of the "ops" field.  Currently, the following segfaults
```
ignore (Obj.new_block 255 1000); Gc.full_major()
```
and now it will not.  I'm still wondering whether it makes any sense at all to create a custom block via `Obj.new_block`, since it's almost impossible to set its "ops" field from OCaml.  The alternative here would be to reject (invalid argument) the case `tg == Custom_tag` and document the limitation.

**Review by xavierleroy — APPROVED** [review]
Looks good to me.  Two remarks below that can be taken into account or ignored.

File: runtime/obj.c (hunk 3 above)
Comment (kayceesrk) [reply runtime/obj.c]: That makes sense. Changed to `0`.

File: runtime/obj.c (hunk 4 above)
Comment (kayceesrk) [reply runtime/obj.c]: I'm not convinced that we should reject custom block allocations. Here is an example of initializing custom blocks from OCaml using a block of custom operations allocated in C.

```c
/* cstub.c */
#include "caml/mlvalues.h"
#include "caml/custom.h"

int compare (value v1, value v2)
{
  return Field(v1,1) - Field(v2,1);
}

static struct custom_operations custom_ops = {
  "_default_ops",
  custom_finalize_default,
  &compare,
  custom_hash_default,
  custom_serialize_default,
  custom_deserialize_default,
  custom_compare_ext_default,
  custom_fixed_length_default
};

value get_custom_ops (value unit)
{
  return (value)&custom_ops;
}
```

```ocaml
/* test.ml */
external get_custom_ops : unit -> Obj.t = "get_custom_ops"

let custom_ops = get_custom_ops ()

let mk_custom_obj v =
  let c = Obj.new_block 255 2 in
  Obj.set_field c 0 custom_ops;
  Obj.set_field c 1 v;
  c

let c1 = mk_custom_obj (Obj.repr 0)
let c2 = mk_custom_obj (Obj.repr 1)

let _ =
  if c1 = c1 then print_endline "c1 = c1";
  if c2 = c2 then print_endline "c2 = c2";
  if c1 <> c2 then print_endline "c1 <> c2"
``` 

```bash
$ ocamlopt.opt test.ml cstub.c
$ ./a.out
c1 = c1
c2 = c2
c1 <> c2
```

There might be better ways to write this. But I can't convince myself to reject this code.

Comment (gasche) [reply runtime/obj.c]: This is only tangentially related to this PR, but have been wishing for a while that we had an OCaml-side API to alloc custom block values. Even if custom block operations remain in C-land (because it's the easiest way to guarantee that they don't rely on the runtime system in unreasonable ways?), it would be nice to be able to create new custom values from OCaml. (Preferably with a proper API rather than `set_field`.)

Comment (xavierleroy) [reply runtime/obj.c]: I thought about that kind of code and others before writing my message, but they all have problems.  In your example, `get_custom_ops` returns a naked pointer, and the first `Obj.set_field` replaces a naked pointer by another.  This will be outlawed pretty soon.  Also, since you must write a C primitive anyway, it would be safer, and even nicer to read, to do the allocation on the C side.

Comment (xavierleroy) [reply runtime/obj.c]: Another bizarre thing that happens when a custom block is created with `Obj.new_block` instead of with `caml_alloc_custom`: if the block is small enough to be allocated in the minor heap, it will never be finalized, even if the ops are changed later to a `struct custom_operation` that has a finalization function.  (Whether to record minor heap blocks for finalization later is decided at allocation time.)

Comment (kayceesrk) [reply runtime/obj.c]: Agreed. The latter behavior is quite bad. I can't see a safe use for `Custom_tag` objects allocated through new_block. I'm happy to reject this case and make the change as you had originally proposed. Given that we're changing the semantics of `new_block`, documenting this in `Changes` is useful. I'll make these changes now.

Comment (kayceesrk) [reply runtime/obj.c]: Done.

**Review by xavierleroy — APPROVED** [review]
Looks fine to me, thanks.  I updated the Changes entry to mark it as a breaking change, which it is in a minor way.  Will merge next.

**Comment by kayceesrk** [comment]
Thanks!
